replicate_ev

kylebmetrum opened the pull request:
@dpastoor Another way to build up a data set from event objects.  Needed this for some project work and wrapped it up into function.  


``` r
library(mrgsolve)
```

    ## mrgsolve: Community Edition

    ## www.github.com/metrumresearchgroup/mrgsolve

``` r
e1 <- ev(amt=100,ii=12,addl=22)
e2 <- ev(amt=100) + ev(amt=200, time=24) + ev(amt=300,time=48)
e3 <- ev(amt=1000) + ev(amt=2000,time=72,rate=1)
```

``` r
idata <- data.frame(ID=1:25)
idata$which_ev <- sample(c("Z", "z","a"),nrow(idata),replace=TRUE)
```

``` r
head(idata)
```

    ##   ID which_ev
    ## 1  1        Z
    ## 2  2        Z
    ## 3  3        Z
    ## 4  4        a
    ## 5  5        Z
    ## 6  6        z

Take in a list of event objects and assign to people in an idata set by a grouping variable

``` r
data <- replicate_ev(list(e1,e3,e2),idata,"which_ev",join=TRUE)
head(data)
```

    ##   time cmt  amt ii addl evid rate ID
    ## 1    0   1  100 12   22    1    0  1
    ## 2    0   1  100 12   22    1    0  2
    ## 3    0   1  100 12   22    1    0  3
    ## 4    0   1 1000  0    0    1    0  4
    ## 5   72   1 2000  0    0    1    1  4
    ## 6    0   1  100 12   22    1    0  5

``` r
mod <- mrgsolve:::house() %>% omat(dmat(0.1,0.1,0.1,0.1))
```

``` r
out <- mod %>% data_set(data) %>% mrgsim(carry.out="evid,cmt,amt,rate")
```

``` r
#plot(out,CP~time|factor(ID),layout=c(5,5))
```

The implementation here scales well (&gt; 200k rows in idata)

``` r
idata <- data.frame(ID=1:200000)
idata$which_ev <- sample(c("Z", "z","a"),nrow(idata),replace=TRUE)
head(idata)
```

    ##   ID which_ev
    ## 1  1        z
    ## 2  2        Z
    ## 3  3        Z
    ## 4  4        z
    ## 5  5        Z
    ## 6  6        Z

``` r
system.time({
  data <- replicate_ev(list(e1,e3,e2),idata,"which_ev")
})
```

    ##    user  system elapsed 
    ##   0.601   0.050   0.653

``` r
dim(data)
```

    ## [1] 400533      8

``` r
head(data)
```

    ##   time cmt  amt ii addl evid rate ID
    ## 1    0   1  100 12   22    1    0  1
    ## 2    0   1 1000  0    0    1    0  2
    ## 3   72   1 2000  0    0    1    1  2
    ## 4    0   1 1000  0    0    1    0  3
    ## 5   72   1 2000  0    0    1    1  3
    ## 6    0   1  100 12   22    1    0  4

``` r
tail(data)
```

    ##        time cmt  amt ii addl evid rate     ID
    ## 400528   72   1 2000  0    0    1    1 199997
    ## 400529    0   1 1000  0    0    1    0 199998
    ## 400530   72   1 2000  0    0    1    1 199998
    ## 400531    0   1 1000  0    0    1    0 199999
    ## 400532   72   1 2000  0    0    1    1 199999
    ## 400533    0   1  100 12   22    1    0 200000

dpastoor commented:
Nice, I see, took me a second to wrap my head around the Z's. How do you handle if the event scenarios don't match up the groupings? 

The only 'problem' I have with this, is intuitively when I see replicate_x, I think of it being applied to each individual, so if you had pointed to this example, without the code results I would have expected ID 1 to have 3 'occasions', each containing said event.

Would it be a more reasonable name to call it like `join_ev`? I might be missing some part of it though?

dpastoor commented on rdev/R/events.R:
```diff
@@ -304,9 +304,82 @@ add.ev <- function(e1,e2) {
+  
+  if(join) {
+    nu <- sapply(idata,is.numeric)
+    x <- dplyr::left_join(x,idata[,nu,drop=FALSE],by="ID") 
```
ok so this is what I was thinking, based on your example, was what was going down. My thoughts on handling the groupings though could be as such:

```
ev_df <- ev_list %>% bind_rows ## list of events but now a df for joining
full_df <- full_join(idata_set, ev_df) # joins based on common grouping var, in example case which_ev
if (any(is.na(full_df$amt)) {
  stop("not all groups were provided dosing events......<additional info here>")
}
```

dpastoor replied:
that could be pretty handy to call in general

kylebmetrum replied:
Sorry for the crappy example; I was testing different input and didn't get it cleaned up to share.

Agree that the naming isn't good.  How about `assign_ev` because the point it that you already have "people" realized and randomized to some intervention and you need to match the intervention with the "person" based on the grouping variable.

I will try to formalize some of the helpers for these events:  enforcing only event-type data items (need to be selective about this) and filling in (1) missing items that have a default (like `evid=1`) and also filling in missing items in one event object because that item appears in another event object and we need to be able to bind them all up.  One option for this is `dplyr::bind_rows` and then just replace `NA` with `0`.  But for an application like this, `bind_rows` is much slower than just `do.call(rbind,list)` when you have a list of matrices and a lot of people (my intended use).

kylebmetrum commented:
Another rider on this PR:

- `mrgsolve:::details` has option (`complete`) to fill in missing parameters and compartments (that weren't annotated)
-  `mrgsolve:::details` has option (`values`)  to add current values for parameters and initial conditions
- Also, if no annotations were provided, `mrgsolve:::details` will return a data frame with all parameters and compartments

kylebmetrum commented:
@dpastoor I also fixed a bug where there were issues if users included more than one sets of parens in an annotation ... for example

```r
[PARAM] @annotated
TVCL : 5 : Typical clearance (in adults) (L/hr)
```

I modified this so that it is the __LAST__ item in parens that gets assigned to `unit`; the remaining parens stay in the item description.  Does that make sense?

dpastoor commented:
nice - that makes complete sense to me

kylebmetrum commented on rdev/R/modspec.R:
```diff
@@ -925,14 +925,14 @@ wrap_namespace <- function(x,name) {
-  what <- dplyr::filter(annot, name %in% capture & block=="PARAM")
+  what <- dplyr::filter(annot, name %in% .capture & block=="PARAM")
   what <- dplyr::mutate(what, block="CAPTURE")
   bind_rows(annot,what)
```
No; no other alias that I know of; but I've been trying to code like that for safety.  I can take it out if it's overkill.

kylebmetrum replied:
And `R CMD CHECK` will force me to import those functions from `dplyr` I believe (?)

dpastoor replied:
yeah you can import them. I like the 'defensive' coding as well, thus might want to add `dplyr::bind_rows` on 937

kylebmetrum commented:
If you capture a parameter but you didn't annotate it, pull from the parameter annotation.  Otherwise, if you do annotate it, that's what you get.  This turned out to be simpler implementation and less fooling around with how stuff is parsed.  

A big change with this is now you can have multiple `$CAPTURE` blocks.

``` r
code <- '
$PARAM @annotated
CL : 1 : Clearance (l/hr/kg)
VC : 2 : Volume (l)
KA : 3 : Absorption rate constant (1/hr)

$CAPTURE @annotated 
DV : Dependent variable
KA : I want this descr!

$CAPTURE CL VC KA dv

$TABLE 
double DV = 1;
double dv = 2;
'

mod <- mcode("capture", code)
```

    ## Compiling capture ... done.

``` r
mrgsolve:::details(mod)
```

    ##     block name                    descr    unit options
    ## 1   PARAM   CL                Clearance l/hr/kg       .
    ## 2   PARAM   VC                   Volume       l       .
    ## 3   PARAM   KA Absorption rate constant    1/hr       .
    ## 4 CAPTURE   DV       Dependent variable       .       .
    ## 5 CAPTURE   KA       I want this descr!       .       .
    ## 6 CAPTURE   CL                Clearance l/hr/kg       .
    ## 7 CAPTURE   VC                   Volume       l       .

dpastoor commented:
Awesome! What happens when you also have a param with no prior annotation
in regards to the details?

On Tue, Nov 22, 2016, 1:06 PM Kyle Baron <notifications@github.com> wrote:

> If you capture a parameter but you didn't annotate it, pull from the
> parameter annotation. Otherwise, if you do annotate it, that's what you
> get. This turned out to be simpler implementation and less fooling around
> with how stuff is parsed.
>
> A big change with this is now you can have multiple $CAPTURE blocks.
>
> code <- '$PARAM @annotatedCL : 1 : Clearance (l/hr/kg)VC : 2 : Volume (l)KA : 3 : Absorption rate constant (1/hr)$CAPTURE @annotated DV : Dependent variableKA : I want this descr!$CAPTURE CL VC KA dv$TABLE double DV = 1;double dv = 2;'
> mod <- mcode("capture", code)
>
> ## Compiling capture ... done.
>
> mrgsolve:::details(mod)
>
> ##     block name                    descr    unit options
> ## 1   PARAM   CL                Clearance l/hr/kg       .
> ## 2   PARAM   VC                   Volume       l       .
> ## 3   PARAM   KA Absorption rate constant    1/hr       .
> ## 4 CAPTURE   DV       Dependent variable       .       .
> ## 5 CAPTURE   KA       I want this descr!       .       .
> ## 6 CAPTURE   CL                Clearance l/hr/kg       .
> ## 7 CAPTURE   VC                   Volume       l       .
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/metrumresearchgroup/mrgsolve/pull/164#issuecomment-262317791>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ADDFmfH8eLK4L4qrup_l7qULMaLCHv-Mks5rAy7jgaJpZM4K0FA8>
> .
>

kylebmetrum commented:
It shows up if you request `complete` listing:

``` r
code <- '
$PARAM LAG = 1.25

$PARAM @annotated
CL : 1 : Clearance (l/hr/kg)
VC : 2 : Volume (l)
KA : 3 : Absorption rate constant (1/hr)

$CAPTURE @annotated 
DV : Dependent variable
KA : I want this descr!

$CAPTURE CL VC KA dv LAG

$TABLE 
double DV = 1;
double dv = 2;
'

mod <- mcode("capture", code)
```

    ## Compiling capture ... done.

``` r
mrgsolve:::details(mod)
```

    ##     block name                    descr    unit options
    ## 1   PARAM   CL                Clearance l/hr/kg       .
    ## 2   PARAM   VC                   Volume       l       .
    ## 3   PARAM   KA Absorption rate constant    1/hr       .
    ## 4 CAPTURE   DV       Dependent variable       .       .
    ## 5 CAPTURE   KA       I want this descr!       .       .
    ## 6 CAPTURE   CL                Clearance l/hr/kg       .
    ## 7 CAPTURE   VC                   Volume       l       .

``` r
mrgsolve:::details(mod,complete=TRUE,value=TRUE)
```

    ##     block name                    descr    unit options value
    ## 1   PARAM   CL                Clearance l/hr/kg       .  1.00
    ## 2   PARAM   VC                   Volume       l       .  2.00
    ## 3   PARAM   KA Absorption rate constant    1/hr       .  3.00
    ## 4 CAPTURE   DV       Dependent variable       .       .    NA
    ## 5 CAPTURE   KA       I want this descr!       .       .  3.00
    ## 6 CAPTURE   CL                Clearance l/hr/kg       .  1.00
    ## 7 CAPTURE   VC                   Volume       l       .  2.00
    ## 8   PARAM  LAG                        .       .       .  1.25

kylebmetrum commented:
A more robust example:


``` r
code <- '
$PARAM LAG = 1.25, CAP = 2.1

$PARAM @annotated
CL : 1 : Clearance (l/hr/kg)
VC : 2 : Volume (l)
KA : 3 : Absorption rate constant (1/hr)

$CAPTURE @annotated 
DV : Dependent variable
KA : I want this descr!
CAP : A captured parameter

$CAPTURE CL VC KA dv LAG

$TABLE 
double DV = 1;
double dv = 2;
'

mod <- mcode("capture", code)
```

    ## Compiling capture ... done.

``` r
mrgsolve:::details(mod)
```

    ##     block name                    descr    unit options
    ## 1   PARAM   CL                Clearance l/hr/kg       .
    ## 2   PARAM   VC                   Volume       l       .
    ## 3   PARAM   KA Absorption rate constant    1/hr       .
    ## 4 CAPTURE   DV       Dependent variable       .       .
    ## 5 CAPTURE   KA       I want this descr!       .       .
    ## 6 CAPTURE  CAP     A captured parameter       .       .
    ## 7 CAPTURE   CL                Clearance l/hr/kg       .
    ## 8 CAPTURE   VC                   Volume       l       .

``` r
mrgsolve:::details(mod,complete=TRUE,value=TRUE)
```

    ##      block name                    descr    unit options value
    ## 1    PARAM   CL                Clearance l/hr/kg       .  1.00
    ## 2    PARAM   VC                   Volume       l       .  2.00
    ## 3    PARAM   KA Absorption rate constant    1/hr       .  3.00
    ## 4  CAPTURE   DV       Dependent variable       .       .    NA
    ## 5  CAPTURE   KA       I want this descr!       .       .  3.00
    ## 6  CAPTURE  CAP     A captured parameter       .       .  2.10
    ## 7  CAPTURE   CL                Clearance l/hr/kg       .  1.00
    ## 8  CAPTURE   VC                   Volume       l       .  2.00
    ## 9    PARAM  LAG                        .       .       .  1.25
    ## 10   PARAM  CAP                        .       .       .  2.10
